apis/v1alpha1/clusterreferencegrant_types.go

```diff
@@ -32,9 +32,33 @@ type ClusterReferenceGrant struct {
 	metav1.TypeMeta   `json:",inline"`
 	metav1.ObjectMeta `json:"metadata,omitempty"`
 
-	From GroupVersionResourcePath `json:"from"`
-	To   GroupResource            `json:"to"`
-	For  For                      `json:"for"`
+	// From refers to the group and resource that these references originate
+	// from.
+	From GroupResource `json:"from"`
+
+	// Versions describes how references and class partitions are defined for
+	// the "From" API. Each Version string must be unique.
+	Versions []VersionedReferencePaths `json:"versions"`
+}
+
+type VersionedReferencePaths struct {
+	Version    string          `json:"version"`
+	References []ReferencePath `json:"references"`
+}
+
+type ReferencePath struct {
+	// Path in the "From" API where referenced names come from.
+	Path string `json:"path"`
+
+	// GroupResource for the target names from the Path
+	To GroupResource `json:"to"`
+
+	// For refers to the purpose of this reference. Subjects of
+	// ClusterReferenceConsumers will be authorized to follow references
+	// matching the From, To, and For of this resource.
+	//
+	// This value must be a valid DNS label as defined per RFC-1035.
```

Comment:
"valid DNS label" is confusing imho. My understanding is that it has nothing to do with DNS but it is will just follow the DNS formatting as defined in RFC-1035. 

Maybe we should change it to "A valid label formatted in a manner akin to DNS conventions as defined in RFC-1035"